### PR TITLE
[DEV-3296] Program tools windowing

### PR DESCRIPTION
**Pull request by ianhitchcock**
#### program-boilerplate
- Added getUserCustomFieldsFromJsonata util to return custom fields being used by programs
- Added meetEdgeTriggerConditions util to check edge conditions of AFTER_USER_CREATED_OR_UPDATED trigger

#### program-test-suite
- remove peer-dependancies
- add test support for edge condition trigger tests

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation or Development tools (readme, specs, tests, code formatting)

## Links

- Jira issue number: DEV-3296
- Process.st launch checklist: https://app.process.st/runs/DEV-3296-Program-custom-field-key-introspection-qSdkUEaNPa37_PjjkHtCXQ/tasks/ipJPSwLUytFe3xMt6P5PYw

### Development

- [ ] [Prettier](https://www.npmjs.com/package/prettier) was run (if applicable)
- [ ] The behaviour changes in the pull request are covered by specs
- [ ] All tests related to the changed code pass in development

### Paperwork

- [x] This pull request has a descriptive title and information useful to a reviewer
- [x] This pull request has a Jira number
- [x] This pull request has a Process.st launch checklist

### Code review

- [ ] Changes have been reviewed by at least one other engineer
- [ ] Security impacts of this change have been considered

### CODE REVIEW

File: packages/program-boilerplate/src/conversion.ts [hunk 1]
```diff
@@ -66,6 +69,47 @@ function parseValue(value: string): RuleValue {
...
+  fields: string[] | undefined,
+  activeTrigger: any
+): boolean {
+  if (fields === undefined || fields.length === 0) {
+    return true;
+  }
+
+  for (const field of fields) {
+    if (!field.startsWith("user.")) {
+      // TODO: what to do here? this is probably some kind of error case
```
Comment (johanventer) [inline packages/program-boilerplate/src/conversion.ts:91]: @jayden-chan I think the right thing to do here might be to fail fast and return `false`. There's obviously some kind of condition trying to be attempted, but we can't process it so denying the trigger feels like the right thing to do.

Please make sure there is a unit test for this case.

File: packages/program-boilerplate/src/conversion.ts [hunk 2]
```diff
@@ -66,6 +69,47 @@ function parseValue(value: string): RuleValue {
...
+      continue;
+    }
+
+    const previousValue = jsonata(field.replace("user.", "previous.")).evaluate(
+      activeTrigger
+    );
+    const currentValue = jsonata(field).evaluate(activeTrigger);
```
Comment (johanventer) [inline packages/program-boilerplate/src/conversion.ts:98]: @jayden-chan These calls to jsonata could fail, you might want to wrap them in try/catch.

Also, while it's unlikely, there's a possibility to cause a denial of service by passing a badly behaved expression. For example, something like `user.blah or ($f := function() { $f() }; $f())` would actually cause an infinite loop.

For this reason please use `safeJsonata`.

File: packages/program-boilerplate/src/utils.ts [hunk 3]
```diff
@@ -1,7 +1,7 @@
 import { rewardScheduleQuery } from "./queries";
 import { ProgramTriggerBody, TriggerType } from "./types/rpc";
-import { User } from "./types/saasquatch";
-import { loggers } from "winston";
+import jsonata = require("jsonata");
```
Comment (johanventer) [inline packages/program-boilerplate/src/utils.ts:3]: @ianhitchcock This import is not necessary, please remove.

**Review by johanventer — CHANGES_REQUESTED** [review]
@ianhitchcock @jayden-chan Some comments inline for recommended changes.

File: packages/program-boilerplate/src/conversion.ts (hunk 2 above)
Comment (jayden-chan) [reply packages/program-boilerplate/src/conversion.ts]: @johanventer I refactored this to use the safe JSONata, but I had to refactor the function signature and it's not backward compatible with the old `safeJsonata`. Previously it was not possible to determine the difference between a successful JSONata that returned `undefined` and a failed JSONata. I need to know this difference for the edge trigger stuff so I just made a new function temporarily. I guess we need to decide if I should just leave the second function in there or if we should increment the major version of program boilerplate.

File: packages/program-boilerplate/src/jsonata.ts [hunk 4]
```diff
@@ -64,3 +65,25 @@ export function safeJsonata(expression: string, inputData: any) {
     logger.warn(`Failed to evaluate JSONata expression: ${e.message}`);
   }
 }
+
+export function safeJsonata2(
```
Comment (johanventer) [inline packages/program-boilerplate/src/jsonata.ts:69]: I would prefer if `safeJsonata` just got new optional arguments with defaults, rather than `safeJsonata2`.

**Review by johanventer — CHANGES_REQUESTED** [review]

File: packages/program-boilerplate/src/jsonata.ts (hunk 4 above)
Comment (jayden-chan) [reply packages/program-boilerplate/src/jsonata.ts:69]: Please see my comment earlier in this PR. The problem is not with the arguments, it's with the return type signature.

File: packages/program-boilerplate/src/conversion.ts (hunk 2 above)
Comment (johanventer) [reply packages/program-boilerplate/src/conversion.ts]: Thanks @jayden-chan I missed this comment, so my other comment about `safeJsonata2` was based on false assumptions.

File: packages/program-boilerplate/src/jsonata.ts (hunk 4 above)
Comment (johanventer) [reply packages/program-boilerplate/src/jsonata.ts:69]: Thanks @jayden-chan, I missed your earlier comment. Could we at least rename the function to something better than `safeJsonata2`?

**Review by johanventer — COMMENTED** [review]
Looks good. Needs final CHANGELOG date and package publishing before merging.

**Review by johanventer — APPROVED** [review]
